### PR TITLE
New: Added branching by attempt and correctness

### LINKED ISSUE #49: Allow branching on attempt number and correctness
### Use-cases

* If the block was correct on the first attempt it should be possible to give extra reward
* If the block was correct beyond the x attempt, the reward should be appropriately changed
* If the block was incorrect on the x attempt, more encouragement should be given

### PR DESCRIPTION
**Pull request by oliverfoster**
fixes #49 

### New
* `_hasAttemptBands` and `_attemptBands`, similar to [`assessmentResults._bands`](https://github.com/adaptlearning/adapt-contrib-assessmentResults/blob/4d506e465dd31ff7549ea24561d6efb078f89a11/example.json#L25-L47), such that branching can occur according to both the block attempt number ranges and block correctness states
* `_useQuestionAttempts` to allow historic completions of a question to determine the attempt branching. Loops can be made across the same question and branch according to the attempt.

See [example.json](https://github.com/adaptlearning/adapt-contrib-branching/blob/899fc48fc0e3481a76674cd9bb51faf4928d9992/example.json#L25-L58)

### CODE REVIEW

**Comment by swashbuck** [comment]
@oliverfoster In the _README.md_, will you remove the version number and add `<br>` line breaks after the other lines (ex. Author / maintainer)?

File: js/branchingAlgorithm.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  const attemptsLeft = questionModels.reduce((sum, questionModel) => sum + (questionModel.get('_attemptsLeft') || questionModel.get('_attempts')), 0);
+  const attemptsTaken = (attemptsPossible - attemptsLeft);
+  return attemptsTaken;
+}
+
+export function getNextId(model) {
+  const config = model.get('_branching');
+  if (config._force) return config._force;
+  const correctness = getCorrectness(model);
+  const hasAttemptBands = Boolean(config?._hasAttemptBands ?? false);
```
Comment (cahirodoherty-learningpool) [inline js/branchingAlgorithm.js:29]: An early return could save us from calculating attemptBand when not applicable
```suggestion
  const hasAttemptBands = Boolean(config?._hasAttemptBands ?? false);
  if (!hasAttemptBands) return config[`_${correctness}`];
```

**Review by zubairslamdien — COMMENTED** [review]
Tested and no issues for me.

**Comment by swashbuck** [comment]
@oliverfoster Can you confirm that this correctly describes how branching works? If so, I can add it to the _README.md_.

> All blocks that are part of a branching sequence need to have a `_branching` object, even if it's empty. For instance, a block can simply use `"_branching": {}` if it should conditionally be shown but does not create any branches of its own.

I'm finding that _not_ including a `_branching` object will cause a block to _always_ appear, even when it's specified in another block as a branch (e.g. `_partlyCorrect`) and even when it belongs to an article that uses branching. I think this note would clarify how it's supposed to work.

**Comment by swashbuck** [comment]
@oliverfoster Is there a need to include `_partlyCorrect` and `_incorrect` in the `_attemptBands` items? The branched block that appears after the question will only appear once the question is _complete_. That is, either the person answers correctly on _any_ attempt or they exhaust _all_ of their attempts.

For example, if a question's number of attempts are set to 10, a branched block would never be shown for an incorrect attempt at attempt #3 since the question isn't complete yet. It would always have to show the "incorrect" block after the 10th attempt. Therefore, there's no reason to specify what happens when the user answers incorrectly on the 3rd or 5th or 7th attempt.

Also, I'm running into some issues in general (could be my JSON configuration). I will DM you a zip of my testing course.

**Comment by oliverfoster** [comment]
> Tested and no issues for me.

Where and how did you test this?

**Comment by oliverfoster** [comment]
> @oliverfoster Can you confirm that this correctly describes how branching works? If so, I can add it to the _README.md_.
> 
> > All blocks that are part of a branching sequence need to have a `_branching` object, even if it's empty. For instance, a block can simply use `"_branching": {}` if it should conditionally be shown but does not create any branches of its own.
> 
> I'm finding that _not_ including a `_branching` object will cause a block to _always_ appear, even when it's specified in another block as a branch (e.g. `_partlyCorrect`) and even when it belongs to an article that uses branching. I think this note would clarify how it's supposed to work.

Correct, thanks.

**Comment by oliverfoster** [comment]
> @oliverfoster Is there a need to include `_partlyCorrect` and `_incorrect` in the `_attemptBands` items? The branched block that appears after the question will only appear once the question is _complete_. That is, either the person answers correctly on _any_ attempt or they exhaust _all_ of their attempts.
> 
> For example, if a question's number of attempts are set to 10, a branched block would never be shown for an incorrect answer on the 3rd attempt since the question isn't complete yet. It would always have to show the "incorrect" block after the 10th attempt. Therefore, there's no reason to specify what happens when the user answers incorrectly on the 3rd or 5th or 7th attempt.
> 
> Also, I'm running into some issues in general (could be my JSON configuration). I will DM you a zip of my testing course.

The block may have two questions, their correctness combines to make block correctness. So it's a little less straightforward than it seems.

It must be article+block branching, not block+component, as we can have unlimited blocks in each article, whereas there a limited components per block in the AAT.

Blocks must therefore combine their components attempts and correctness.

If the block only had one question then your use-case stands - partlyCorrect and incorrect would not be needed until the final attempt, all previous attempts would only complete on correct responses.

**Comment by zubairslamdien** [comment]
> > Tested and no issues for me.
> 
> Where and how did you test this?

I just set branching for an MCQ with two correct answers and added 2 attempts as a band, but I'll test more.

**Comment by chris-steele** [comment]
Would it be correct to state that this new feature gives us _attempt-based-branching_ rather than _attempt-based-feedback_? The latter being different question feedback depending on which attempt at the question you are on.

**Comment by oliverfoster** [comment]
`_useQuestionAttempts`  - Controls whether this block will branch according previous completed attempts at this question

branching question attempts

**Review by StuartNicholls — APPROVED** [review]

**Comment by StuartNicholls** [comment]
The only thing that through me off a little was having to set 

`"_shouldStoreAttempts": true`

I see you have put a check in for this though. Just missed at first as it fails 'silently'.

Couldn't find any issues however

**Comment by oliverfoster** [comment]
Fails silently by posting an error to the console:
https://github.com/adaptlearning/adapt-contrib-branching/blob/5c17b66bc666471388ac2f39df57fbf93120b53a/js/adapt-contrib-branching.js#L43

**Comment by StuartNicholls** [comment]
> Fails silently by posting an error to the console:
> 
> https://github.com/adaptlearning/adapt-contrib-branching/blob/5c17b66bc666471388ac2f39df57fbf93120b53a/js/adapt-contrib-branching.js#L43

Yep, all good, I did see that

**Review by chris-steele — APPROVED** [review]

**Pull request merged**